packages/xod-project/src/patch.js

```diff
@@ -887,3 +887,123 @@ export const removeDebugNodes = def(
     listNodes,
   )(patch)
 );
+
+/**
+ * Returns a map of pins for patch that exists.
+ * So Pins are fully valid, contains proper types and etc.
+ */
+export const getPinsForExistentPatch = def(
+  'getPinsForExistentPatch :: Node -> Patch -> Map PinKey Pin',
+  (node, patch) => R.compose(
+    R.map(pin => R.assoc(
+      'value',
+      Node.getBoundValueOrDefault(pin, node),
+      pin
+    )),
+    patchPins => R.compose(
+      R.mergeWith(R.merge, R.__, patchPins),
+      R.map(R.compose(
+        R.objOf('normalizedLabel'),
+        Pin.getPinLabel
+      )),
+      R.indexBy(Pin.getPinKey),
+      Pin.normalizePinLabels,
+      R.values
+    )(patchPins),
+    R.indexBy(Pin.getPinKey),
+    listPins
+  )(patch)
+);
+
+/**
+ * Returns a map of pins for Node, that points to unexistent Patch.
+ * These pins are created from links, connected with some Pins in this Node.
+ */
+export const getPinsForUnexistentPatch = def(
+  'getPinsForUnexistentPatch :: Node -> Patch -> Map PinKey Pin',
+  (node, currentPatch) => {
```

Comment:
The same as above: `listDeadNodePins`.

Not obvious why the second arg is called “current” patch.

BTW, “unexsistent” is exotic… “nonexisting” is casual.